diff --git a/src/HAL9000/src/system.c b/src/HAL9000/src/system.c
index 48aac9e..6cf5bcc 100644
--- a/src/HAL9000/src/system.c
+++ b/src/HAL9000/src/system.c
@@ -70,7 +70,7 @@ SystemInit(
     status = STATUS_SUCCESS;
     pCpu = NULL;
 
-    LogSystemInit(LogLevelError,
+    LogSystemInit(LogLevelInfo,
                   LogComponentInterrupt | LogComponentIo | LogComponentAcpi,
                   TRUE
                   );
